File: epi_judge_python_solutions/run_length_compression.py
import itertools

from test_framework import generic_test
from test_framework.test_failure import TestFailure


def decoding(s):

    count, result = 0, []
    for c in s:
        if c.isdigit():
            count = count * 10 + int(c)
        else:  # c is a letter of alphabet.
            result.append(c * count)  # Appends count copies of c to result.
            count = 0
    return ''.join(result)


def encoding(s):

    result, count = [], 1
    for i in range(1, len(s) + 1):
        if i == len(s) or s[i] != s[i - 1]:
            # Found new character so write the count of previous character.
            result.append(str(count) + s[i - 1])
            count = 1
        else:  # s[i] == s[i - 1].
            count += 1
    return ''.join(result)


def encoding_pythonic(s):
    return ''.join(str(len(list(g))) + c for c, g in itertools.groupby(s))


def rle_tester(encoded, decoded):
    if decoding(encoded) != decoded:
        raise TestFailure('Decoding failed')
    if encoding(decoded) != encoded:
        raise TestFailure('Encoding failed')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("run_length_compression.py",
                                       'run_length_compression.tsv',
                                       rle_tester))

File: epi_judge_cpp_solutions/run_length_compression.cc
#include <cctype>
#include <string>

#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"

using std::string;
using std::to_string;

string Decoding(const string &s) {
  int count = 0;
  string result;
  for (const char &c : s) {
    if (isdigit(c)) {
      count = count * 10 + c - '0';
    } else {                    // c is a letter of alphabet.
      result.append(count, c);  // Appends count copies of c to result.
      count = 0;
    }
  }
  return result;
}

string Encoding(const string &s) {
  string result;
  for (int i = 1, count = 1; i <= size(s); ++i) {
    if (i == size(s) || s[i] != s[i - 1]) {
      // Found new character so write the count of previous character.
      result += to_string(count) + s[i - 1];
      count = 1;
    } else {  // s[i] == s[i - 1].
      ++count;
    }
  }
  return result;
}

void RleTester(const string &encoded, const string &decoded) {
  if (Decoding(encoded) != decoded) {
    throw TestFailure("Decoding failed");
  }
  if (Encoding(decoded) != encoded) {
    throw TestFailure("Encoding failed");
  }
}

// clang-format off


int main(int argc, char* argv[]) {
  std::vector<std::string> args {argv + 1, argv + argc};
  std::vector<std::string> param_names {"encoded", "decoded"};
  return GenericTestMain(args, "run_length_compression.cc", "run_length_compression.tsv", &RleTester,
                         DefaultComparator{}, param_names);
}
// clang-format on
